experimental/oak_async/src/grpc.rs

```diff
@@ -129,3 +145,116 @@ impl<T> Drop for MultiWriter<T> {
         }
     }
 }
+
+/// A streaming gRPC client request.
+///
+/// `T` represents the message type sent in the requests.
+pub struct GrpcRequestStream<T: Decodable> {
+    inner: Peekable<ChannelReadStream<GrpcRequest>>,
+    is_drained: bool,
+    _msg: core::marker::PhantomData<T>,
+}
+
+// Unpin is needed to be able to peek at the first request before passing it
+// on to the handler.
+impl<T: Decodable> Unpin for GrpcRequestStream<T> {}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
+impl GrpcRequestStream<()> {
+    pub fn new(inner: ChannelReadStream<GrpcRequest>) -> GrpcRequestStream<()> {
+        GrpcRequestStream {
+            inner: inner.peekable(),
+            is_drained: false,
+            _msg: core::marker::PhantomData,
+        }
+    }
+
+    // Casts this stream into the given message type.
+    pub fn into_type<T: Decodable>(self) -> GrpcRequestStream<T> {
+        GrpcRequestStream {
+            inner: self.inner,
+            is_drained: self.is_drained,
+            _msg: core::marker::PhantomData,
+        }
+    }
+}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
+impl<T: Decodable> GrpcRequestStream<T> {
+    pub fn peek(self: Pin<&mut Self>) -> Peek<'_, ChannelReadStream<GrpcRequest>> {
+        Peekable::peek(Pin::new(&mut Pin::into_inner(self).inner))
+    }
+
+    pub async fn first(mut self) -> Result<T, OakError> {
+        // Unwrapping is safe here, as the gRPC stream must have at least one incoming request
+        self.next().await.unwrap()
+    }
+}
+
+impl<T: Decodable> Stream for GrpcRequestStream<T> {
+    type Item = Result<T, OakError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.is_drained {
+            return Poll::Ready(None);
+        }
+        let self_ = Pin::into_inner(self);
+        let inner = Pin::new(&mut self_.inner);
+        Peekable::poll_next(inner, cx).map(|ready| {
+            ready.map(|some| {
+                some.and_then(|request| {
+                    self_.is_drained = request.last;
+                    let message = Message {
+                        bytes: request.req_msg,
+                        handles: Vec::new(),
+                    };
+                    T::decode(&message)
+                })
+            })
+        })
+    }
+}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
```

Comment:
Adding a new struct, with specific field names, for the return type might be overkill, but I think it is still useful to have a bit of documentation here, explaining what this function produces.

Reply:
I think it's a good idea, done!